WcaOnRails/app/controllers/delegate_reports_controller.rb

```diff
@@ -0,0 +1,56 @@
+class DelegateReportsController < ApplicationController
+  before_action :authenticate_user!
+
+  before_action -> { redirect_unless_user(:can_view_delegate_report?, competition_from_params) }, only: [
+    :show,
+  ]
+  before_action -> { redirect_unless_user(:can_edit_competition_report?, competition_from_params) }, only: [
+    :edit,
+    :update,
+  ]
+
+  private def competition_from_params
+    if params[:competition_id]
+      competition = Competition.find(params[:competition_id])
+    else
+      report = DelegateReport.find(params[:id])
+      competition = report.competition
+    end
+    competition
+  end
+
+  def show
+    @competition = competition_from_params
+    @delegate_report = @competition.delegate_report
+  end
+
+  def edit
+    @competition = competition_from_params
+    @delegate_report = @competition.delegate_report
+  end
+
+  def update
+    @competition = competition_from_params
+    @delegate_report = @competition.delegate_report
+    was_posted = @delegate_report.posted?
+    if @delegate_report.update_attributes(delegate_report_params)
+      flash[:success] = "Updated report"
+      if @delegate_report.posted?
+        if !was_posted
+          CompetitionsMailer.notify_of_delegate_report_submission(@competition).deliver_later
+          flash[:info] = "Your report has been submitted and emailed to reports@worldcubeassociation.org!"
+        end
+        redirect_to competition_report_path(@competition)
+      else
+        redirect_to competition_report_edit_path(@competition)
+      end
+    else
+      flash.now[:danger] = "Could not update the report!"
```

Comment:
I fear that for long forms, the problematic fields won't necessarily show up on the screen, and you might think the submission *did* succeed.

I hear you, though. Perhaps the solution is to be consistent about our use of `shared/error_messages`? Perhaps we could get SimpleForm to render it automatically? My other paranoia is that `update_attributes` might fail, but there are no validation errors to show the user. I don't have enough experience with Rails to know if that is possible. Perhaps our `shared/error_messages` could become smart enough to check if the object is not persisted, not new, *and* has no validation errors, and render a scary looking "oh-oh" message??

I've created #674 to discuss this, please continue the discussion over there.

In the mean time, I'll remove this flash message, and add a call to `render 'shared/error_messages'`.